src/azul/service/responseobjects/storage_service.py

```diff
@@ -37,14 +41,156 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
 
     def create_bucket(self, bucket_name: str = None):
         self.client.create_bucket(Bucket=(bucket_name or self.__bucket_name))
-        logger.warning(f'{type(self).__name__}: Created a bucket called "{bucket_name or self.__bucket_name}"')
+        # logger.warning(f'{type(self).__name__}: Created a bucket called "{bucket_name or self.__bucket_name}"')
+
+
+class MultipartUploadHandler:
+
+    def __init__(self, bucket_name, object_key, upload_id):
+        self.__resource = boto3.resource('s3')
+        self.__bucket_name = bucket_name
+        self.__object_key = object_key
+        self.__upload_id = upload_id
+        self.__handler = self.__resource.MultipartUpload(self.__bucket_name, self.__object_key, self.__upload_id)
+        self.__next_part_number = 1
+        self.__closed = False
+        self.__parts = []
+        self.__upload_lock = Lock()
+
+    @property
+    def is_active(self):
+        return not self.__closed
+
+    def complete(self):
+        if not self.is_active:
+            return
+
+        if not self.__parts:
+            # logger.warning(f'multipart_upload: the upload is empty.')
+            self.abort()
+            # logger.warning(f'multipart_upload: aborted')
+            raise EmptyMultipartUploadError(f'{self.__bucket_name}/{self.__object_key}')
+
+        # logger.info(f'multipart_upload: Waiting for active uploads to complete.')
+        while self._has_active_uploads():
+            pass  # Ensure that hanging threads
+        # logger.info(f'multipart_upload: Check the last part.')
+
+        last_part = self.__parts[-1]
+
+        if not last_part.already_uploaded:
+            # Per documentation, the last part can be at any size. Hence, the uploadable condition is ignored.
+            # logger.warning(f'multipart_upload: uploading the last part')
+            self._upload_part(last_part)
+            # logger.info(f'multipart_upload: uploaded the last part')
+
+        self.__handler.complete(MultipartUpload={"Parts": [part.to_dict() for part in self.__parts]})
+        self.__handler = None
+        self.__closed = True
+
+    def abort(self):
+        if not self.is_active:
+            return
+
+        self.__handler.abort()
+        self.__handler = None
+        self.__closed = True
+
+    def push(self, data: bytes):
+        # If the last part is under the minimum limit, the data will be appended.
+        if not self.__parts:
+            part = self._create_new_part(data)
+        else:
+            last_part = self.__parts[-1]
+            if last_part.already_uploaded:
+                part = self._create_new_part(data)
+            else:
+                part = last_part
+                part.content.append(data)
+
+        if not part.is_uploadable:
+            return
+
+        # logger.info(f'multipart_upload/push: Part {part.part_number}: uploading in a background thread')
+        Thread(target=self._upload_part, args=(part,)).start()
```

Comment:
Starting threads like this, and not joining them is a deadly sin.

Nowadays you shouldn't have to use threads directly anymore. It's almost always more advisable to use a ThreadPoolExecutor.